### PR TITLE
Allow configuration of snapshot column names

### LINKED ISSUE #10185: Allow customizing names of metadata fields in snapshots
From https://github.com/dbt-labs/dbt-core/issues/9848

### Description

There are a handful of [Snapshot meta fields](https://docs.getdbt.com/docs/build/snapshots#snapshot-meta-fields):
* `dbt_valid_from`
* `dbt_valid_to`
* `dbt_scd_id`
* `dbt_updated_at`

These are fields dbt creates when creating the snapshot.

Folks want the ability to rename these meta fields - “I want to rename `dbt_valid_from` to `grace_valid_from`”- because of some corporate standard

Customers currently have to create a view on top of the snapshot
    
```sql
select 
    dbt_valid_to as grace_valid_to,
    ...
from {{ ref('my_snapshot') }}
```

We should instead provide the ability to rename these meta fields within the actual snapshot configuration.

### draft of YML config:

```yaml
{{
    config(
      target_database='analytics',
      target_schema='snapshots',
      unique_key='id',

      strategy='timestamp',
      updated_at='updated_at',
      
      dbt_field_name_overrides={
              dbt_valid_from: grace_valid_from,
              dbt_valid_to: grace_valid_to
      }
    )
}}
```

Open questions/thoughts:
- Ideas on name of config: `snapshot_field_names`, `meta_field_aliases`, `snapshot_aliases`, `meta_field_names`, `snapshot_meta_fields`, `snapshot_meta_columns`, `snapshot_columns_names`, `snapshot_dbt_field_name_overrides`
- You should also be able to set a default for this at the project level
    
```yaml
snapshots:
  my_project:
    +dbt_field_name_overrides={
      dbt_valid_from: grace_valid_from,
      dbt_valid_to: grace_valid_to
    }
```   
- If I have an existing snapshot, and I want to rename the fields (add this config)… what should happen?
    - [**let's go with this to start**] option 1: throw an error
        - the error message should indicate that they’ve have to go in and manually update their existing snapshots with the new column names
    - option 2: map old names → new names
    - option 3: don’t touch the snapshots, automatically create a view on top of it
        - this breaks our "1:1" relationship from snapshot to warehouse object :((
    - option 4: some sort of codegen that codegens the view for them
    - option 5: execute the rename for them
        - but we’d need both the old name (old config) and the new name
        - not possible :(
    - option 6: we create new fields, and coalesce with old?
        - same problems as option 5

### Acceptance Critera

- new config for snapshots to alias / override the names of the metadata fields `dbt_field_name_overrides`
- can set that config for individual snapshot, or project default
- throw an error if they have an existing snapshot and they try to add/change this config - this should error on insert if you change the meta field names, but [one idea] we could try to throw an error before running the snapshot DML by running `adapters.get_columns_in_relation()` on the existing snapshot and comparing the meta fields that exist in the current config to what exists in the warehouse 
  - Create a follow-up issue for migration.

### PR DESCRIPTION
**Pull request by gshank**
Resolves #10185

### Problem

We hardcode the column names in snapshot tables.

### Solution

Allow configuring column names.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [x] I have run this code in development, and it appears to resolve the stated issue.
- [x] This PR includes tests, or tests are not required or relevant for this PR.
- [x] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

### CODE REVIEW

File: .changes/unreleased/Features-20240903-154133.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+kind: Features
+body: Allow configuring snapshot column names
```
Comment (graciegoheen) [inline .changes/unreleased/Features-20240903-154133.yaml:2]: I think there are duplicate changelogs?

Comment (gshank) [reply .changes/unreleased/Features-20240903-154133.yaml:2]: Yep. I removed one of them.

File: dev-requirements.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
-git+https://github.com/dbt-labs/dbt-adapters.git@main
-git+https://github.com/dbt-labs/dbt-adapters.git@main#subdirectory=dbt-tests-adapter
+git+https://github.com/dbt-labs/dbt-adapters.git@snapshot_column_names
```
Comment (peterallenwebb) [inline dev-requirements.txt:1]: I know it's obvious, but I'm just leaving a reminder that these changes will need to be left out of the merge.

**Review by peterallenwebb — APPROVED** [review]
Looks good!

**Pull request merged**